src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Pbkdf2Implementation.Unix.cs

```diff
@@ -0,0 +1,32 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Security.Cryptography;
+
+namespace Internal.Cryptography
+{
+    internal partial class Pbkdf2Implementation
+    {
+        public static unsafe void Fill(
+            ReadOnlySpan<byte> password,
+            ReadOnlySpan<byte> salt,
+            int iterations,
+            HashAlgorithmName hashAlgorithmName,
+            Span<byte> destination)
+        {
+            Debug.Assert(!destination.IsEmpty);
+            Debug.Assert(hashAlgorithmName.Name is not null);
+            IntPtr evpHashType = HashProviderDispenser.HashAlgorithmToEvp(hashAlgorithmName.Name);
+            int result = Interop.Crypto.Pbkdf2(password, salt, iterations, evpHashType, destination);
+            const int Success = 1;
+
+            if (result != Success)
+            {
+                Debug.Assert(result == 0, $"Unexpected result {result}");
```

Comment:
> Are internal implementation failures potentially expected; e.g., OOM?

Unclear exactly what you are asking, but I will attempt to elaborate:

1. The actual OpenSSL function `PKCS5_PBKDF2_HMAC` will return 0 on failure, including if it cannot allocate, and 1 on success.
2. We wrap the OpenSSL function with a shim that returns whatever OpenSSL returns. However, we also return `-1` if any of the inputs are valid, like a `NULL` destination or something. That should never happen since we should be validating that the inputs are sane prior to passing to the shim, the shim is just being defensive.
3. So the check here is saying, if it wasn't successful, it better be because the return is "0" (an error) as opposed to "oh no we somehow let bad inputs get to the shim.

All that said, I think then `Debug.Assert` is the right thing to do here, it's a "we have a problematic code path that get bad inputs get to the shim".

This pattern is fairly common throughout the OpenSSL shims.

Reply:
> So the check here is saying, if it wasn't successful, it better be because the return is "0" (an error) as opposed to "oh no we somehow let bad inputs get to the shim".

This clarifies things - thanks!